### PR TITLE
Fixed issue with onErrorDropped being called when using concatWith in QuorumReader

### PR DESCRIPTION
**Pull request by kushagraThapar**
* `concatWith` operator in QuorumReader is not used correctly, and the error surfaced when reactor core fixed the behavior of FluxConcatArray in this PR - https://github.com/reactor/reactor-core/pull/2742 in the version 3.4.9 and above. 
* Fixed behavior by using switchIfEmpty() which is more appropriate in the logic. 
* Added comments to understand the logic better. 
* Fixes https://github.com/Azure/azure-sdk-for-java/issues/24125

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/QuorumReader.java [hunk 1]
```diff
@@ -242,7 +242,9 @@ public QuorumReader(
                     });
             }).repeat(maxNumberOfReadQuorumRetries)
                    .takeUntil(dummy -> !shouldRetryOnSecondary.v)
-                .concatWith(Flux.defer(() -> {
+                   //   In case there is an empty response from above flatMap, it means we could not complete read quorum
+                   //   So we will throw an error, which will be eventually retried.
+                   .switchIfEmpty(Flux.defer(() -> {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/QuorumReader.java:247]: this was a bug, thanks for fixing it @kushagraThapar

**Review by moderakh — APPROVED** [review]
Thanks @kushagraThapar nice catch on the bug.  Thanks for fixing it.

LGTM.

**Review by simplynaveen20 — APPROVED** [review]
LGTM Thanks

**Review by srnagar — APPROVED** [review]

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks

**Pull request merged**